flink-state-backends/flink-statebackend-rocksdb/src/test/java/org/apache/flink/contrib/streaming/state/RocksIncrementalCheckpointRescalingBenchmarkTest.java

```diff
@@ -0,0 +1,240 @@
+            for (int i = 0; i < partitionParallelism; i++) {
+                // Initialize, open, and then re-snapshot the two subtasks to create a partitioned
+                // incremental RocksDB snapshot.
+                OperatorSubtaskState subtaskState =
+                        AbstractStreamOperatorTestHarness.repartitionOperatorState(
+                                snapshot, maxParallelism, 1, partitionParallelism, i);
+                KeyGroupRange localKeyGroupRange20 = keyGroupPartitions.get(i);
+
+                partitionedTestHarness[i] =
+                        getHarnessTest(keySelector, maxParallelism, partitionParallelism, i);
+                partitionedTestHarness[i].setStateBackend(getStateBackend());
+                partitionedTestHarness[i].setup();
+                partitionedTestHarness[i].initializeState(subtaskState);
+                partitionedTestHarness[i].open();
+            }
+
+            partitionedSnapshot =
+                    AbstractStreamOperatorTestHarness.repackageState(
+                            partitionedTestHarness[0].snapshot(1, 2),
+                            partitionedTestHarness[1].snapshot(1, 2));
+
+        } finally {
+            closeHarness(partitionedTestHarness);
+        }
+    }
+
+    @Test(timeout = 1000)
+    @RetryOnFailure(times = 3)
+    public void benchmarkScalingUp() throws Exception {
+        long benchmarkTime = 0;
+
+        // Trigger the incremental re-scaling via restoreWithRescaling by repartitioning it from
+        // parallelism of >1 to a higher parallelism. Time spent during this step includes the cost
+        // of incremental rescaling.
+        List<KeyGroupRange> keyGroupPartitions =
+                StateAssignmentOperation.createKeyGroupPartitions(
+                        maxParallelism, repartitionParallelism);
+
+        long fullStateSize = partitionedSnapshot.getStateSize();
+
+        for (int i = 0; i < repartitionParallelism; i++) {
+            OperatorSubtaskState subtaskState =
+                    AbstractStreamOperatorTestHarness.repartitionOperatorState(
+                            partitionedSnapshot,
+                            maxParallelism,
+                            partitionParallelism,
+                            repartitionParallelism,
+                            i);
+            KeyGroupRange localKeyGroupRange20 = keyGroupPartitions.get(i);
+
+            try (KeyedOneInputStreamOperatorTestHarness<Integer, Integer, Integer> subtaskHarness =
+                    getHarnessTest(keySelector, maxParallelism, repartitionParallelism, i)) {
+                RocksDBStateBackend backend = getStateBackend();
+                subtaskHarness.setStateBackend(backend);
+                subtaskHarness.setup();
+
+                // Precisely measure the call-site that triggers the restore operation.
+                long startingTime = System.nanoTime();
+                subtaskHarness.initializeState(subtaskState);
```

Comment:
`restoreWithScaling` is called through this method but I was unclear if there were successive operations that would have caused a difference in performance.

On my commit with the larger benchmark workload I had also tried measuring the time of the entire process and didn't see other areas with increased/decreased performance -- but with this many records and on a local laptop, the methodology was not very scientific.